### PR TITLE
Fix #861 Introduce wizard-like help system

### LINKED ISSUE #861: Introduce wizard-like help system
We should introduce a wizard like help system taking extra care that it will work nicely on the mobile where it will be the default help system.

### PR DESCRIPTION
**Pull request by allyoucanmap**
* added react-joyride library to controll the steps of tutorial
* added tutorial plugin | autostart or start from burger menu tutorial button
* added new ids to some components to target with joyride
* added tests for the new actions and reducer

main options in local configuration ("name": "Tutorial"):
- showCheckbox: true/false, shows or not the checkbox on intro to disable the tutorial autostart
- rawSteps: array of step objects

Step with intro selector represents the introduction and if it exists must be the first of rawSteps.
If there isn't introduction, auto start is disabled and the tutorial can be activated from the burger menu.

standard step object:

```
{
  "translation": "intro", // add a reference in the traslation file
  "selector": "intro", // could be a presetted selector (plugin/tutorial/selectors.js) or a custom id (#my-id)
  "position": "top" // position of the popup || "bottom" default
}
```

in translation file

```
"tutorial": {
   ... ,
  "intro": {
    "title": "Welcome on MapStore2",
    "text": "My description"
  }
}
```

### CODE REVIEW

File: web/client/localConfig.json [hunk 1]
```diff
@@ -100,6 +100,29 @@
           }
         }
     }, {
+        "name": "Tutorial",
+        "cfg": {
+          "rawSteps": [
```
Comment (offtherailz) [inline web/client/localConfig.json]: This should be not here by default, only if you want to change the steps

File: web/client/localConfig.json [hunk 2]
```diff
@@ -100,6 +100,29 @@
...
+            {
+              "translation": "intro",
+              "selector": "intro"
```
Comment (offtherailz) [inline web/client/localConfig.json]: You should place here the selector, and not a reference to it

File: web/client/localConfig.json [hunk 3]
```diff
@@ -100,6 +100,29 @@
...
+            },
+            {
+              "translation": "lorem",
```
Comment (offtherailz) [inline web/client/localConfig.json]: In the default configuration we should already have some minimal introduction, we can customize and extend later, not a lorem ipsum

File: web/client/localConfig.json [hunk 4]
```diff
@@ -326,6 +349,33 @@
           }
         }
     }, {
+        "name": "Tutorial",
+        "cfg": {
+          "rawSteps": [
```
Comment (offtherailz) [inline web/client/localConfig.json]: same here, remove cfg, use defaults for localConfig

File: web/client/localConfig.json [hunk 5]
```diff
@@ -787,7 +837,38 @@
       {"name": "MapType"},
       {"name": "CreateNewMap"},
       {"name": "Maps"},
-      {"name": "Examples"}
+      {"name": "Examples"},
+      {
```
Comment (offtherailz) [inline web/client/localConfig.json]: If you want 2 different default configurations for home and map, use something a "preset": <home>|<map> you can easily override with `localConfig.json`.
You can save that preset as separate files in the plugin folders. rawSteps configuration will override the steps in the preset.

File: web/client/plugins/Tutorial.jsx [hunk 6]
```diff
@@ -0,0 +1,221 @@
...
+const defaultStep = {
+    title: '',
+    text: '',
+    selector: '',
+    position: 'bottom',
+    type: 'click',
+    allowClicksThruHole: true
+};
+
+const TutorialComponent = React.createClass({
```
Comment (offtherailz) [inline web/client/plugins/Tutorial.jsx]: This should be a separate component. To test.

File: web/client/plugins/tutorial/selectors.js [hunk 7]
```diff
@@ -0,0 +1,31 @@
+/**
+ * Copyright 2017, GeoSolutions Sas.
+ * All rights reserved.
+ *
+ * This source code is licensed under the BSD-style license found in the
+ * LICENSE file in the root directory of this source tree.
+ */
+
+const selectors = {
```
Comment (offtherailz) [inline web/client/plugins/tutorial/selectors.js]: Remove this file, the selectors goes directly in the rawSteps array. Instead, create presets for home and map to use. 
So in your localConfig.json you will have something like` {name: "Tutorial", preset: "home"}` or ` {name: "Tutorial", preset: "map"}`

Comment (allyoucanmap) [inline web/client/plugins/tutorial/selectors.js]: Could it be like this? Or we always have to declare the preset

preset home
```
{
 "name": "Tutorial",
 "cfg": {
  "preset": "home"
 }
}
```

preset map 

```
{
 "name": "Tutorial"
}
```

Comment (offtherailz) [inline web/client/plugins/tutorial/selectors.js]: Yes, good

**Comment by allyoucanmap** [comment]
there are plugin that disappear or change their structure based on events.
an example is the search bar

error0

error1

this doesn't permit to the library react-joyride to mount the popup to the target,
so I added a div with the id "error-tutorial" as target for every step with error.
When the user close the tutorial, every error detected is deleted.

error2

error3

File: web/client/actions/tutorial.js [hunk 8]
```diff
@@ -0,0 +1,175 @@
...
+function setupTutorial(introSteps) {
+    return (dispatch, getState) => {
+        var state = getState();
+        var steps = assign([], introSteps);
+        var hasIntro = false;
+        var disabled = localStorage.getItem('mapstore.plugin.tutorial.disabled');
+        if (state.controls && state.controls.tutorial) {
+            dispatch(setControlProperty('tutorial', 'enabled', false));
+        }
+        steps.map((step) => {
```
Comment (mbarto) [inline web/client/actions/tutorial.js]: do not use map with side effects

Comment (mbarto) [inline web/client/actions/tutorial.js]: either use forEach or (better) reduce

File: web/client/actions/tutorial.js [hunk 9]
```diff
@@ -0,0 +1,175 @@
...
+function changeStatusTutorial(status) {
+    return {
+        type: CHANGE_STATUS_TUTORIAL,
+        status: status
+    };
+}
+
+function setupTutorial(introSteps) {
+    return (dispatch, getState) => {
+        var state = getState();
```
Comment (mbarto) [inline web/client/actions/tutorial.js]: Please, never use var, either use const (preferred) or let

File: web/client/actions/tutorial.js [hunk 10]
```diff
@@ -0,0 +1,175 @@
...
+
+function changeStatusTutorial(status) {
+    return {
+        type: CHANGE_STATUS_TUTORIAL,
+        status: status
+    };
+}
+
+function setupTutorial(introSteps) {
+    return (dispatch, getState) => {
```
Comment (mbarto) [inline web/client/actions/tutorial.js]: Using getState in actions is usually evil: reducers are the place where we manipulate state

File: web/client/actions/tutorial.js [hunk 11]
```diff
@@ -0,0 +1,175 @@
...
+        return step.selector !== '#intro-tutorial';
+    });
+    dispatch(setStepsTutorial(steps));
+    if (tour.action !== 'next') {
+        dispatch(closeTutorial());
+    }
+}
+
+function updateTutorial(tour, updateSteps, error) {
+    return (dispatch, getState) => {
```
Comment (mbarto) [inline web/client/actions/tutorial.js]: This is a reducer, not an action

File: web/client/actions/tutorial.js [hunk 12]
```diff
@@ -0,0 +1,175 @@
...
+            dispatch(changeStatusTutorial(tour.type));
+        }
+    };
+}
+
+function disableTutorial() {
+    return (dispatch, getState) => {
+        dispatch(toggleTutorial());
+        let state = getState();
+        let disabled = state && state.tutorial && state.tutorial.disabled;
```
Comment (mbarto) [inline web/client/actions/tutorial.js]: This is a reducer, not an action

File: web/client/actions/tutorial.js [hunk 13]
```diff
@@ -0,0 +1,175 @@
...
+        localStorage.setItem('mapstore.plugin.tutorial.disabled', disabled);
+    };
+}
+
+function resetTutorial() {
+    return (dispatch, getState) => {
```
Comment (mbarto) [inline web/client/actions/tutorial.js]: This is a reducer, not an action

File: web/client/components/tutorial/Tutorial.jsx [hunk 14]
```diff
@@ -0,0 +1,190 @@
...
+                onSetup: () => {},
+                onStart: () => {},
+                onUpdate: () => {},
+                onToggle: () => {},
+                onReset: () => {}
+            }
+        };
+    },
+    componentWillMount() {
+        let rawSteps = this.props.rawSteps.length > 0 ? this.props.rawSteps : this.props.presetList[this.props.preset] || [];
```
Comment (mbarto) [inline web/client/components/tutorial/Tutorial.jsx]: We try to not use lifecycle methods, if not necessary. This is state initialization code, I presume so should be developed as actions / reducers code

File: web/client/components/tutorial/style/style.js [hunk 15]
```diff
@@ -0,0 +1,149 @@
+/**
+ * Copyright 2017, GeoSolutions Sas.
+ * All rights reserved.
+ *
+ * This source code is licensed under the BSD-style license found in the
+ * LICENSE file in the root directory of this source tree.
+ */
+
+const introStyle = {
```
Comment (mbarto) [inline web/client/components/tutorial/style/style.js]: Can this be just css?

Comment (allyoucanmap) [inline web/client/components/tutorial/style/style.js]: We could use css to override the default style but we still need the style to change the tooltips of intro and error. Could we remove the style.js file and include the const introStyle and errorStyle in the Tutorial.jsx component?

**Comment by offtherailz** [comment]
Looks very good. We should disable the intro until we have a stable version to show up. Please document the preset format.

**Pull request merged**